### PR TITLE
Feature/40009

### PR DESCRIPTION
**Pull request by sa11616079**
Create a file authMoiddleWare.ts inside libs/routes.

Install a dependency 'jsonwebtoken',

Create a middleware 'authMiddleWare' with following:

Accepts two arguments(module and permission type).
Retrieve token from the header 'Authorization' and decode it using .verify method. It will give the details of user.
If no user decoded. Throws an error of 403. Unauthorized Access.
Use the function hasPermission defined earlier in utils for checking the permission.
If permission does not exists, throws error.
 

Sample token for various roles are:

eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJTdWNjZXNzaXZlIFRlY2hub2xvZ2llcyIsImlhdCI6MTU0ODY5MDQzNywiZXhwIjoxNTgwNzk3MjU5LCJhdWQiOiJ3d3cuc3VjY2Vzc2l2ZS5pbiIsInN1YiI6IkxlYXJuIGFuZCBJbXBsZW1lbnQiLCJuYW1lIjoiVHJhaW5lZSIsImVtYWlsIjoidHJhaW5lZUBzdWNjZXNzaXZlLnRlY2giLCJyb2xlIjoidHJhaW5lZSIsImp0aSI6ImEwMjBhYTJlLTk1ZDUtNDc5MC05NDAxLWYyMjFlNDBiNjA1MCJ9.-z2lCFkZJaW5ACZurZH2NiAh9VhefjRCvmlDw1X57Rs
{ "iss": "Successive Technologies", "iat": 1548690437, "exp": 1580226437, "aud": "www.successive.in&quot;, "sub": "Learn and Implement", "name": "Trainee", "email": "trainee@successive.tech", "role": "trainee" } key: qwertyuiopasdfghjklzxcvbnm123456
eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJTdWNjZXNzaXZlIFRlY2hub2xvZ2llcyIsImlhdCI6MTU0ODY5MDQzNywiZXhwIjoxNTgwNzk3MzkxLCJhdWQiOiJ3d3cuc3VjY2Vzc2l2ZS5pbiIsInN1YiI6IkxlYXJuIGFuZCBJbXBsZW1lbnQiLCJuYW1lIjoiSGVhZCBUcmFpbmVyIiwiZW1haWwiOiJoZWFkLnRyYWluZXJAc3VjY2Vzc2l2ZS50ZWNoIiwicm9sZSI6ImhlYWQtdHJhaW5lciIsImp0aSI6IjhhMjNkNjc1LWY1NjctNDNmMy1iMDQxLWE3YzlkM2VhMzI4ZCJ9.ciRD8ypgm90pOfD5AD2NKKGOnE7Ke4M0ubSNZvQIaKw { "iss": "Successive Technologies", "iat": 1548690437, "exp": 1580226437, "aud": "www.successive.in&quot;, "sub": "Learn and Implement", "name": "Head Trainer", "email": "head.trainer@successive.tech", "role": "head-trainer" } key: qwertyuiopasdfghjklzxcvbnm123456

### CODE REVIEW

File: src/libs/routes/authMoiddleWare.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+export default (moduleName:string, permissionType:string) => (req:Request, res:Response, next:NextFunction) => {
+    try 
+    {
+        console.log("The config is : ", moduleName, permissionType);
+        console.log("Header is ", req.headers['authorization']);
+        const token = req.headers['authorization'];
+        const decodeUser = jwt.verify(token, 'fQ5JGYpHyISVsBr2OFHHuV1z4cO0nFmL');  
+        const role=decodeUser.role;
+        console.log('User',decodeUser);
+        if(hasPermission(permissions.getUser,role,permissionType))
```
Comment (SHUBHAM-JAIN-SS) [inline src/libs/routes/authMoiddleWare.ts:16]: use modulename dynamically

**Review by SHUBHAM-JAIN-SS — CHANGES_REQUESTED** [review]

File: src/libs/routes/authMoiddleWare.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+export default (moduleName:string, permissionType:string) => (req:Request, res:Response, next:NextFunction) => {
+    try 
+    {
+        console.log("The config is : ", moduleName, permissionType);
+        console.log("Header is ", req.headers['authorization']);
+        const token = req.headers['authorization'];
+        const decodeUser = jwt.verify(token, 'fQ5JGYpHyISVsBr2OFHHuV1z4cO0nFmL');  
+        const role=decodeUser.role;
+        console.log('User',decodeUser);
+        if(hasPermission(moduleName,role,permissionType))
```
Comment (SHUBHAM-JAIN-SS) [inline src/libs/routes/authMoiddleWare.ts:16]: you are passing a string here and expecting an object in your function implemetation

**Review by SHUBHAM-JAIN-SS — CHANGES_REQUESTED** [review]

File: src/controllers/trainee/routes.ts [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
+import authMoiddleWare from "../../libs/routes/authMoiddleWare";
+
 const traineeRouter=new Router();
 
 traineeRouter.route('/')
-    .get(TraineeController.get)
-    .post(TraineeController.create)
-    .put(TraineeController.update)
-    .delete(TraineeController.delete);
+    .get(authMoiddleWare('getUser','read'),validationHandler(Validation.get), TraineeController.get)
```
Comment (SHUBHAM-JAIN-SS) [inline src/controllers/trainee/routes.ts:10]: you are passing a string here and expecting an object in your middleware implementation

**Review by SHUBHAM-JAIN-SS — CHANGES_REQUESTED** [review]

File: src/controllers/trainee/routes.ts (hunk 3 above)
Comment (sa11616079) [reply src/controllers/trainee/routes.ts]: check it sir again

**Pull request merged**